### PR TITLE
feature/add create endpoint

### PR DESCRIPTION
**Pull request by borovHokage**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/865

### CODE REVIEW

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using System.Linq;
+
+namespace LT.DigitalOffice.HistoryService.Business.Commands.Service
+{
+    public class CreateServiceCommand : ICreateServiceCommand
+    {
+        private readonly IServiceRepository _repository;
+        private readonly IDbServiceMapper _mapperService;
+        private readonly IAccessValidator _accessValidator;
+        public readonly ICreateServiceRequestValidator _validator;
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: public?

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            _repository = repository;
+            _mapperService = mapperService;
+            _accessValidator = accessValidator;
+            _validator = validator;
+        }
+
+        public OperationResultResponse<Guid> Execute(CreateServiceRequest request)
+        {
+            if (!(_accessValidator.IsAdmin() || _accessValidator.HasRights(Rights.AddEditRemoveHistroies)))
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: use only hasrights

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            {
+                throw new ForbiddenException("Not enough rights.");
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: don't throw exc

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            if (_repository.IsServiceNameExist(request.Name))
+            {
+                response.Status = OperationResultStatusType.Failed;
+                response.Errors.Add($"Service with name '{request.Name}' already exist");
+                return response;
+            }
+
+            _validator.ValidateAndThrowCustom(request);
+
+            DbService dbService = _mapperService.Map(request);
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: you don't need variable

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+        public FindServiceCommand(
+            IServiceRepository repository,
+            IFindServiceResponseMapper mapper)
+        {
+            _repository = repository;
+            _mapper = mapper;
+        }
+
+        public List<ServiceInfo> Execute()
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: why don't you use OperationResultResponse or Find?

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            List<DbService> dbServiceList = _repository.Find();
+            List<ServiceInfo> response = new List<ServiceInfo>();
```
Comment (BelehovEgor) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: why don't you use select

File: src/HistoryService.Data/ServiceRepository.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        private readonly IDataProvider _provider;
+
+        public ServiceRepository(
+            IDataProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public bool IsServiceNameExist(string name)
```
Comment (BelehovEgor) [inline src/HistoryService.Data/ServiceRepository.cs]: does

File: src/HistoryService.Mappers/Responses/Interfaces/IFindServiceResponseMapper.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using LT.DigitalOffice.HistoryService.Models.Db;
+using LT.DigitalOffice.HistoryService.Models.Dto.Responses;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.HistoryService.Mappers.Responses.Interfaces
+{
+    [AutoInject]
+    public interface IFindServiceResponseMapper
+    {
+        public ServiceInfo Map(DbService dbService);
```
Comment (BelehovEgor) [inline src/HistoryService.Mappers/Responses/Interfaces/IFindServiceResponseMapper.cs]: remove public

Comment (BelehovEgor) [reply src/HistoryService.Mappers/Responses/Interfaces/IFindServiceResponseMapper.cs]: name of mapper doesn't match return type

File: src/HistoryService/HistoryService.csproj [hunk 9]
```diff
@@ -5,6 +5,7 @@
   <ItemGroup>
 	<PackageReference Include="AspNetCore.HealthChecks.SqlServer" Version="5.0.2" />
 	<PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="5.0.1" />
+	<PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="5.0.9" />
```
Comment (BelehovEgor) [inline src/HistoryService/HistoryService.csproj:8]: fix format

File: src/HistoryService/Startup.cs [hunk 10]
```diff
@@ -73,7 +73,7 @@ public Startup(IConfiguration configuration)
                 .GetSection(BaseRabbitMqConfig.SectionName)
                 .Get<BaseRabbitMqConfig>();
 
-            Version = "1.0.0";
+            Version = "1.1.0";
```
Comment (BelehovEgor) [inline src/HistoryService/Startup.cs]: use 4 number for spec

**Review by Cheephot — APPROVED** [review]

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
+﻿using LT.DigitalOffice.HistoryService.Business.Commands.Service.Interfaces;
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:1]: add editor config to project

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            if (!(_accessValidator.IsAdmin() || _accessValidator.HasRights(Rights.AddEditRemoveHistroies)))
+            {
+                throw new ForbiddenException("Not enough rights.");
+            }
+
+            OperationResultResponse<Guid> response = new();
+
+            if (_repository.IsServiceNameExist(request.Name))
+            {
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: add http status code Conflict

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                response.Status = OperationResultStatusType.Failed;
+                response.Errors.Add($"Service with name '{request.Name}' already exist");
+                return response;
+            }
+
+            _validator.ValidateAndThrowCustom(request);
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: use VatidateCustom

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+            DbService dbService = _mapperService.Map(request);
+
+            response.Body = _repository.Create(dbService);
+
+            response.Status = response.Errors.Any() ? OperationResultStatusType.PartialSuccess : OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: errors newer has any() in this place

Comment (MariyaSin) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: check if body == null

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:90]: add http status code Created

File: src/HistoryService.Business/Commands/Service/Interfaces/IFindServiceCommand.cs [hunk 16]
```diff
@@ -0,0 +1,12 @@
+﻿using LT.DigitalOffice.HistoryService.Models.Dto.Responses;
+using LT.DigitalOffice.Kernel.Attributes;
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.HistoryService.Business.Commands.Service.Interfaces
+{
+    [AutoInject]
+    public interface IFindServiceCommand
+    {
+        List<ServiceInfo> Execute();
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/Interfaces/IFindServiceCommand.cs]: FindResultResponse

File: src/HistoryService.Data/ServiceHistoryRepository.cs [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+            IDataProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public Guid Create(DbServiceHistory dbServiceHistory)
+        {
+            if (dbServiceHistory == null)
+            {
+                throw new ArgumentNullException(nameof(dbServiceHistory));
```
Comment (MariyaSin) [inline src/HistoryService.Data/ServiceHistoryRepository.cs]: do not throw exceptions

File: src/HistoryService.Data/ServiceRepository.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            return _provider.Services.Any(p => p.Name.Contains(name));
+        }
+
+        public Guid Create(DbService dbService)
+        {
+            if (dbService == null)
+            {
+                throw new ArgumentNullException(nameof(dbService));
```
Comment (MariyaSin) [inline src/HistoryService.Data/ServiceRepository.cs]: do not throw exceptions

File: src/HistoryService.Mappers/Db/DbServiceMapper.cs [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        public DbServiceMapper(IHttpContextAccessor httpContextAccessor)
+        {
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        public DbService Map(CreateServiceRequest request)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
```
Comment (MariyaSin) [inline src/HistoryService.Mappers/Db/DbServiceMapper.cs]: do not throw exceptions

File: src/HistoryService.Mappers/Responses/FindServiceResponseMapper.cs [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace LT.DigitalOffice.HistoryService.Mappers.Responses
+{
+    public class FindServiceResponseMapper : IFindServiceResponseMapper
+    {
+        public ServiceInfo Map(DbService dbService)
+        {
+            if (dbService == null)
+            {
+                throw new ArgumentNullException(nameof(dbService));
```
Comment (MariyaSin) [inline src/HistoryService.Mappers/Responses/FindServiceResponseMapper.cs]: remove exseption

File: src/HistoryService.Validation/Service/CreateServiceRequestValidator.cs [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+using LT.DigitalOffice.HistoryService.Validation.Service.Interfaces;
+
+namespace LT.DigitalOffice.HistoryService.Validation.Service
+{
+    public class CreateServiceRequestValidator : AbstractValidator<CreateServiceRequest>, ICreateServiceRequestValidator
+    {
+        public CreateServiceRequestValidator()
+        {
+            RuleFor(service => service.Name.Trim())
+                    .NotEmpty()
```
Comment (MariyaSin) [inline src/HistoryService.Validation/Service/CreateServiceRequestValidator.cs]: check tab length

File: src/HistoryService/Controllers/ServiceController.cs [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+        [HttpPost("create")]
+        public OperationResultResponse<Guid> Create(
+            [FromServices] ICreateServiceCommand command,
+            [FromBody] CreateServiceRequest request)
+        {
+            var result = command.Execute(request);
+
+            if (result.Status != OperationResultStatusType.Failed)
+            {
+                _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
```
Comment (MariyaSin) [inline src/HistoryService/Controllers/ServiceController.cs]: remove to command

File: src/HistoryService/Controllers/ServiceController.cs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+            }
+
+            return result;
+        }
+
+        [HttpGet("find")]
+        public List<ServiceInfo> Find(
```
Comment (MariyaSin) [inline src/HistoryService/Controllers/ServiceController.cs]: FinedResultResponse

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 1 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: fixed

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 3 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: fixed

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs (hunk 6 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 4 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: done

File: src/HistoryService.Data/ServiceRepository.cs (hunk 7 above)
Comment (borovHokage) [reply src/HistoryService.Data/ServiceRepository.cs]: fixed

File: src/HistoryService.Mappers/Responses/Interfaces/IFindServiceResponseMapper.cs (hunk 8 above)
Comment (borovHokage) [reply src/HistoryService.Mappers/Responses/Interfaces/IFindServiceResponseMapper.cs]: fixed

File: src/HistoryService/HistoryService.csproj (hunk 9 above)
Comment (borovHokage) [reply src/HistoryService/HistoryService.csproj:8]: resolved

File: src/HistoryService/Startup.cs (hunk 10 above)
Comment (borovHokage) [reply src/HistoryService/Startup.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 2 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 11 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:1]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 13 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 14 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: done

File: src/HistoryService.Business/Commands/Service/Interfaces/IFindServiceCommand.cs (hunk 16 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/Interfaces/IFindServiceCommand.cs]: done

File: src/HistoryService.Data/ServiceRepository.cs (hunk 18 above)
Comment (borovHokage) [reply src/HistoryService.Data/ServiceRepository.cs]: done

File: src/HistoryService.Mappers/Db/DbServiceMapper.cs (hunk 19 above)
Comment (borovHokage) [reply src/HistoryService.Mappers/Db/DbServiceMapper.cs]: done

File: src/HistoryService.Mappers/Responses/FindServiceResponseMapper.cs (hunk 20 above)
Comment (borovHokage) [reply src/HistoryService.Mappers/Responses/FindServiceResponseMapper.cs]: done

File: src/HistoryService/Controllers/ServiceController.cs (hunk 22 above)
Comment (borovHokage) [reply src/HistoryService/Controllers/ServiceController.cs]: done

File: src/HistoryService/Controllers/ServiceController.cs (hunk 23 above)
Comment (borovHokage) [reply src/HistoryService/Controllers/ServiceController.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 15 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:90]: done

File: src/HistoryService.Data/ServiceHistoryRepository.cs (hunk 17 above)
Comment (borovHokage) [reply src/HistoryService.Data/ServiceHistoryRepository.cs]: done

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 12 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs]: done

File: src/HistoryService.Validation/Service/CreateServiceRequestValidator.cs (hunk 21 above)
Comment (borovHokage) [reply src/HistoryService.Validation/Service/CreateServiceRequestValidator.cs]: fixed

File: src/HistoryService/Controllers/ServiceController.cs [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+using System.Collections.Generic;
+using System.Net;
+
+namespace LT.DigitalOffice.HistoryService.Controllers
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class ServiceController : ControllerBase
+    {
+        private readonly IHttpContextAccessor _httpContextAccessor;
```
Comment (MariyaSin) [inline src/HistoryService/Controllers/ServiceController.cs]: you do not use it there

Comment (borovHokage) [reply src/HistoryService/Controllers/ServiceController.cs]: fixed

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+        return new OperationResultResponse<Guid?>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = errors
+        };
+      }
+
+      if (_repository.DoesServiceNameExist(request.Name))
+      {
+        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:66]: add empty line

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            _repository = repository;
+            _mapper = mapper;
+            _httpContextAccessor = httpContextAccessor;
+           _accessValidator = accessValidator;
+        }
+
+        public OperationResultResponse<List<FindResultResponse>> Execute()
+        {
+
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: remove empty line

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+                return new OperationResultResponse<List<FindResultResponse>>
+                {
+                  Status = OperationResultStatusType.Failed,
+                  Errors = new() { "Not enough rights." }
+                };
+            }
+            OperationResultResponse<List<FindResultResponse>> response = new();
+
+            response.Body = _repository.Find().Select(x => _mapper.Map(x)).ToList();
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: you can do it shortest:  .....Select(_mapper.Map).ToList();

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+            if (response.Body == null)
+            {
+                _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
+
+                response.Status = OperationResultStatusType.Failed;
+                response.Errors.Add("Services are not exist");
+                return response;
+            }
+
+            response.Status = OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: remove to 51 line? and delete 58 string

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+            OperationResultResponse<List<FindResultResponse>> response = new();
+
+            response.Body = _repository.Find().Select(x => _mapper.Map(x)).ToList();
+
+            if (response.Body == null)
+            {
+                _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
+
+                response.Status = OperationResultStatusType.Failed;
+                response.Errors.Add("Services are not exist");
```
Comment (MariyaSin) [inline src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: no need in this error, we return errors then try to get by id

File: src/HistoryService.Models.Dto/Responses/FindResultResponse.cs [hunk 30]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace LT.DigitalOffice.HistoryService.Models.Dto.Responses
+{
+    public record FindResultResponse
```
Comment (MariyaSin) [inline src/HistoryService.Models.Dto/Responses/FindResultResponse.cs]: rename by entity name.

File: src/HistoryService/Controllers/ServiceController.cs [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+using LT.DigitalOffice.Kernel.Responses;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Collections.Generic;
+using System.Net;
+
+namespace LT.DigitalOffice.HistoryService.Controllers
+{
+    [Route("[controller]")]
```
Comment (MariyaSin) [inline src/HistoryService/Controllers/ServiceController.cs]: is it works?

File: src/HistoryService/Controllers/ServiceController.cs [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+    [ApiController]
+    public class ServiceController : ControllerBase
+    {
+        [HttpPost("create")]
+        public OperationResultResponse<Guid?> Create(
+            [FromServices] ICreateServiceCommand command,
+            [FromBody] CreateServiceRequest request)
+        {
+            var result = command.Execute(request);
```
Comment (MariyaSin) [inline src/HistoryService/Controllers/ServiceController.cs]: for what do you create the variable?

File: src/HistoryService/Controllers/ServiceController.cs (hunk 31 above)
Comment (BelehovEgor) [reply src/HistoryService/Controllers/ServiceController.cs]: what is your problem @MariyaSin?

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs (hunk 28 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: fixed

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs (hunk 29 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: fixed

File: src/HistoryService.Models.Dto/Responses/FindResultResponse.cs (hunk 30 above)
Comment (borovHokage) [reply src/HistoryService.Models.Dto/Responses/FindResultResponse.cs]: fixed

File: src/HistoryService/Controllers/ServiceController.cs (hunk 32 above)
Comment (borovHokage) [reply src/HistoryService/Controllers/ServiceController.cs]: delete

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs (hunk 27 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: fixed

File: src/HistoryService.Business/Commands/Service/FindServiceCommand.cs (hunk 26 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/FindServiceCommand.cs]: fixed

File: src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs (hunk 25 above)
Comment (borovHokage) [reply src/HistoryService.Business/Commands/Service/CreateServiceCommand.cs:66]: done

**Comment by borovHokage** [comment]
mistake

**Pull request closed without merging**